### PR TITLE
Plasma guns

### PR DESCRIPTION
**Pull request by TiviPlus**
## About The Pull Request
Added six new plasma guns: [Plasma pistol, Plasma rifle, Plasma carbine, Plasma minigun, Plasma cannon, Plasma sniper]
They have their own firemodes and have overheat mechanics

## Changelog
:cl:
add: Added the plasma series of weapons
/:cl:

### CODE REVIEW

**Comment by FosterGrime-Heymynameslouie** [comment]
THE PLASMA MACHINE GUN IS HERE

**Comment by Mrrpip** [comment]
To save our mother earth from any alien attack!

**Comment by Naaanii** [comment]
Where niche?

**Comment by 1kajowsmol** [comment]
> Where niche?

joe niche

**Review by Kuro020 — APPROVED** [review]

**Comment by DeltaFire15** [comment]
"code/modules/projectiles/gun_system.dm:		overheat_timer = addtimer(CALLBACK(src, .proc/complete_overheat, overheat_smoke), overheat_time, TIMER_STOPPABLE)
ERROR: Outdated proc reference use detected in code, please use proc reference helpers.
"

**Comment by Lumipharon** [comment]
Overheat seems wack.

Many firemodes will not overheat even if you m1 the entire mag. Other modes like the basic rifle overheat super fast. And again, I don't see how anything stops marines just juggling two plasma guns to bypass overheat.

Trifire is bugged, it fires all projectiles from the firer.

**Comment by imsxz** [comment]
bouncy goes thru shooter

**Comment by 1kajowsmol** [comment]
currently seems like overheat does basically nothing and also the shotgun mode on the carbine is op as fuck and breaks reinforced walls in like 4 mags, idk about actual damage.

**Comment by Kuro020** [comment]
standard rifle firemode is using the wrong projectile sprite

**Comment by imsxz** [comment]
carbine shotgun burst crits xenos and kills them in under 0.5s from full HP. base damage seems insanely high regardless of bugged burst.

short and neat descriptions that accurately described each of the firing modes would be nice when i examine a gun. dont like the wall of lore personally when im trying to figure out what it does.

**Comment by imsxz** [comment]
the plasma thrower and plasma cannon among other things have  icons that are way too big for your inventory

**Comment by Kuro020** [comment]
the plasma sniper scope doesnt actually do anything, gotta make it a permanent attachment

**Comment by Kuro020** [comment]
image

**Comment by Lumipharon** [comment]
@TiviPlus you probs don't need the flags on every guin if they're the same lol

File: code/modules/projectiles/gun_system.dm [hunk 1]
```diff
@@ -980,6 +1016,18 @@
...
 
+///called by a timer after overheat finishes
+/obj/item/weapon/gun/proc/complete_overheat(overheat_smoke)
+	QDEL_NULL(overheat_smoke)
+	overheat_timer = null
+	heat_amount = 0
+
+/obj/item/weapon/gun/process(delta_time)
+	heat_amount = max(0, heat_amount - cool_amount*delta_time)
+	heat_meter.animate_change(heat_amount/100, 5)
```
Comment (Lumipharon) [inline code/modules/projectiles/gun_system.dm:1027]: this is spamming runtimes with deployed sentries for some reason.

**Review by Lumipharon — CHANGES_REQUESTED** [review]

File: code/modules/projectiles/gun_system.dm (hunk 1 above)
Comment (Lumipharon) [reply code/modules/projectiles/gun_system.dm:1027]: Oh it breaks sentries entirely, apparently

File: code/modules/codex/entries/guns_codex.dm [hunk 2]
```diff
@@ -146,6 +146,16 @@
 		traditional chargers and their capability to switch their lens, allowing more flexibility, something that a ballistic weapon \
 		aren't capable of."
 
+/datum/codex_entry/plasma_weapons
```
Comment (XSlayer300) [inline code/modules/codex/entries/guns_codex.dm:149]: what is this point of adding a codex entry if you haven't listed them yet in the generic plasma weapon code?

File: code/modules/projectiles/guns/plasma.dm [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+	fire_delay = 0.45 SECONDS
+	heat_per_fire = 10
+	rounds_per_shot = 66
+	radial_icon_state = "plasma_strong"
+
+/datum/lasrifle/base/plasma_rifle_mode/carbine_trifire
+	fire_sound = 'sound/weapons/guns/fire/plasma_fire_fast.ogg'
+	ammo_datum_type = /datum/ammo/energy/plasma/carbine_trifire
+	icon_state = "plasma_carbine"
+	fire_mode = GUN_FIREMODE_AUTOMATIC
```
Comment (XSlayer300) [inline code/modules/projectiles/guns/plasma.dm:250]:
```suggestion
	fire_mode = GUN_FIREMODE_AUTOMATIC
```
>tri-fire
>fully-automatic

File: code/modules/projectiles/guns/plasma.dm [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+		/obj/item/weapon/gun/grenade_launcher/underslung,
+		/obj/item/attachable/motiondetector,
+		/obj/item/weapon/gun/rifle/pepperball/pepperball_mini,
+	)
+
+/datum/lasrifle/base/plasma_rifle_mode/carbine_standard
+	fire_sound = 'sound/weapons/guns/fire/plasma_fire_fast.ogg'
+	ammo_datum_type = /datum/ammo/energy/plasma/carbine_standard/four
+	icon_state = "plasma_carbine"
+	fire_mode = GUN_FIREMODE_AUTOBURST
```
Comment (XSlayer300) [inline code/modules/projectiles/guns/plasma.dm:229]: inverted from tri-fire.

File: code/modules/projectiles/guns/plasma.dm [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+	fire_sound = 'sound/weapons/guns/fire/plasma_fire_med.ogg'
+	unload_sound = 'sound/weapons/guns/interact/plasma_unload_1.ogg'
+	reload_sound = 'sound/weapons/guns/interact/plasma_reload_1.ogg'
+	ammo_datum_type = /datum/ammo/energy/plasma/pistol_standard
+	force = 15
+	flags_gun_features = GUN_CAN_POINTBLANK|GUN_AMMO_COUNTER|GUN_AMMO_COUNT_BY_SHOTS_REMAINING
+	mode_list = list(
+		"Standard" = /datum/lasrifle/base/plasma_rifle_mode/pistol_standard,
+		"Automatic" = /datum/lasrifle/base/plasma_rifle_mode/pistol_automatic,
+		"Tri-fire" = /datum/lasrifle/base/plasma_rifle_mode/pistol_trifire,
```
Comment (XSlayer300) [inline code/modules/projectiles/guns/plasma.dm:270]:
```suggestion
		"Tri-fire" = /datum/lasrifle/base/plasma_rifle_mode/pistol_trifire,
```
Name it bouncy mode or something, it's not even three round burst.

File: code/modules/projectiles/guns/plasma.dm [hunk 6]
```diff
@@ -0,0 +1,385 @@
...
+	radial_icon_state = "plasma_cannon"
+
+/datum/lasrifle/base/plasma_rifle_mode/cannon_heavy
+	fire_sound = 'sound/weapons/guns/fire/plasma_fire_blast.ogg'
+	ammo_datum_type = /datum/ammo/energy/plasma/cannon_heavy
+	fire_mode = GUN_FIREMODE_AUTOMATIC
+	fire_delay = 2
+	heat_per_fire = 35
+	rounds_per_shot = 150
+	icon_state = "plasma_cannon"
```
Comment (XSlayer300) [inline code/modules/projectiles/guns/plasma.dm:374]: No icon exists.

**Comment by FosterGrime-Heymynameslouie** [comment]
will there be an ammo box specific for it like the t60 ammo box and t12 ?

**Comment by FosterGrime-Heymynameslouie** [comment]
the scope for the marksman rifle dosent seem to work

also the plasma rifle when being vended out and placing a new mag shows 83 ammo but when changing fire modes to the other 2 will show 53 instead and the overcharge mechanic wont work anymore

**Comment by ivanmixo** [comment]
https://github.com/tgstation/TerraGov-Marine-Corps/assets/22431091/abac9aee-de11-425e-ac7c-d561acde4b06

Please reconsider the whole heavy mode on the cannon, this is not okay.

**Comment by AusarAtkin** [comment]
Plasma Pistol on standard mode kill Primo King and Queen in one mag, it shoot as fast as you can click, and kill most primo t3 and t2 in half a mag, and it wont ever overheat magdumping the whole mag.

**Comment by Kuro020** [comment]
Plasma cannon fire rate is probably on byond tick instead of seconds, should fix that, also give all the guns (aside from the pistol) the two handed required to fire flag

**Comment by carbous** [comment]
So my feedback on those guns, I  didn't try the pistol, canon and flamer yet. The rifle is quite decent, it seems to do what it was meant to, good sunder, decent damages but not too much and the marksman mode is funny. The sniper was quite bad, good sunder but overral less than the other plasma rifle simply because it's could get hard to get a hit in (still better than the laser sniper). The carbine was' damages were alright, not too much, however I think it's ammo size might be a bit too big. I also worry about the sunder dealt, coupled with it's high firerate, it might be a bit too much.

File: code/modules/projectiles/guns/plasma.dm [hunk 7]
```diff
@@ -0,0 +1,385 @@
...
+	muzzleflash_iconstate = "muzzle_flash_pulse"
+
+/obj/item/weapon/gun/energy/lasgun/lasrifle/plasma/minigun
+	name = "\improper PL-73 plasma thrower"
+	desc = "The PL-73 Plasma Thrower is a high-quality weapon designed to deliver powerful performance in even the most challenging combat situations. With its sleek, durable design and intuitive controls, this plasma gun is the ultimate choice for those who demand the very best.\n\nFeaturing three distinct fire modes - rapid, incendiary, and glob - the PL-73 is capable of unleashing a devastating array of plasma-based attacks. Whether you're looking to overwhelm your enemies with a barrage of rapid-fire plasma bolts, ignite the surrounding area with intense heat and flames, or deliver a massive plasma ball that explodes on impact, the PL-73 has you covered.\n\nConstructed from only the highest quality materials, this weapon is built to last. Its ergonomic design and intuitive controls make it easy to handle, even in the heat of battle. So if you're looking for a powerful and reliable plasma gun that can help you take on even the toughest opponents, look no further than the PL-73 Plasma Thrower."
```
Comment (XSlayer300) [inline code/modules/projectiles/guns/plasma.dm:12]:
```suggestion
	desc = "A heavy duty minigun-like plasma-discharging weapon, for those who want to go all out in a superheated siege. Comes with multiple charge settings that either rapid fire plasma bolts, discharge plasma energy into flames, or deliver a massive plasma ball which deals heavy damage. Kongjian's sleek PL series of plasma-discharging weapons acts as a competitor to TE series of laser weapons used by the TGMC, allowing flexibility even in the most stressful combat situations, though the usage of highest quality materials and specifications made large-scale use and procurement all but impossible. Uses proprietary WML plasma batteries, and overheats when used too much, depending on setting."
```

File: code/modules/projectiles/guns/plasma.dm [hunk 8]
```diff
@@ -0,0 +1,385 @@
...
+	ammo_datum_type = /datum/ammo/energy/plasma/minigun_glob
+	heat_per_fire = 25
+	fire_delay = 0.15 SECONDS
+	fire_mode = GUN_FIREMODE_SEMIAUTO
+	rounds_per_shot = 100
+	radial_icon_state = "plasma_cannon"
+
+/obj/item/weapon/gun/energy/lasgun/lasrifle/plasma/sniper
+	name = "\improper PL-02 plasma sniper rifle"
+	desc = "Introducing the PL-02 Plasma Sniper Rifle, the ultimate long-range weapon for those who demand nothing but the best. With its advanced features and sleek design, the PL-02 is the perfect addition to any marksman's arsenal. This powerful plasma rifle is designed for precision and accuracy, allowing you to take out your enemies from a distance with ease. Its advanced targeting system and high-powered scope make it easy to acquire and track targets, while its powerful plasma bolts deliver devastating damage. Constructed from high-quality materials, the PL-02 is built to last and can withstand even the toughest of environments. If you're looking for a reliable and powerful sniper rifle that can take out targets from a distance, look no further than the PL-02 Plasma Sniper Rifle."
```
Comment (XSlayer300) [inline code/modules/projectiles/guns/plasma.dm:85]:
```suggestion
	desc = "Kongjian's answer to the marksman question, the PL-02 isn't as flexible compared to the other TGMC plasma weapons, but it is just as fierce and incredibly accurate as the most weapons in the arsenal. Kongjian's sleek PL series of plasma-discharging weapons acts as a competitor to TE series of laser weapons used by the TGMC, allowing flexibility even in the most stressful combat situations, though the usage of highest quality materials and specifications made large-scale use and procurement all but impossible. Uses proprietary WML plasma batteries, and overheats when used too much, depending on setting."
```

File: code/modules/projectiles/guns/plasma.dm [hunk 9]
```diff
@@ -0,0 +1,385 @@
...
+		/obj/item/weapon/gun/shotgun/combat/masterkey,
+		/obj/item/weapon/gun/flamer/mini_flamer,
+		/obj/item/weapon/gun/grenade_launcher/underslung,
+		/obj/item/attachable/motiondetector,
+		/obj/item/weapon/gun/rifle/pepperball/pepperball_mini,
+	)
+
+/obj/item/weapon/gun/energy/lasgun/lasrifle/plasma/rifle
+	name = "\improper PL-38 plasma rifle"
+	desc = "Introducing the PL-38 Plasma Rifle, a versatile weapon for any combat situation. With its sleek design and advanced features, the PL-38 is the perfect addition to any soldier's arsenal. This powerful plasma rifle is built for speed and accuracy, allowing you to take out your enemies with ease. Its advanced targeting system and high-powered scope make it easy to acquire and track targets, while its marksman and standard firemodes give you the flexibility to adapt to any situation. Constructed from high-quality materials, the PL-38 is built to withstand even the toughest of combat environments. Whether you're on the front lines of battle or engaging in covert operations, the PL-38 Plasma Rifle is the weapon of choice for any soldier looking for a reliable and powerful plasma rifle."
```
Comment (XSlayer300) [inline code/modules/projectiles/guns/plasma.dm:124]:
```suggestion
	desc = "The versatile jack-of-all-trades standard plasma rifle, with a fully automatic charge setting as well as a stronger, 'marksman' mode for the accurate role. Kongjian's sleek PL series of plasma-discharging weapons acts as a competitor to TE series of laser weapons used by the TGMC, allowing flexibility even in the most stressful combat situations, though the usage of highest quality materials and specifications made large-scale use and procurement all but impossible. Uses proprietary WML plasma batteries, and overheats when used too much, depending on setting."
```

File: code/modules/projectiles/guns/plasma.dm [hunk 10]
```diff
@@ -0,0 +1,385 @@
...
+	icon_state = "plasma_rifle"
+	fire_mode = GUN_FIREMODE_AUTOMATIC
+	fire_delay = 0.45 SECONDS
+	heat_per_fire = 2
+	rounds_per_shot = 40
+	radial_icon_state = "plasma_strong"
+
+/obj/item/weapon/gun/energy/lasgun/lasrifle/plasma/carbine
+	name = "\improper PL-51 plasma carbine"
+	desc = "Introducing the PL-51 Plasma Carbine, a compact and lightweight weapon designed for close to mid-range combat. With its sleek design and advanced features, the PL-51 is the perfect addition to any soldier's arsenal. This powerful plasma carbine is built for speed and agility, allowing you to quickly take out your enemies in close quarters combat. Its advanced targeting system and high-powered scope make it easy to acquire and track targets, while its standard, shotgun, and tri-fire firemodes give you the flexibility to adapt to any combat situation. Constructed from high-quality materials, the PL-51 is built to withstand even the toughest of environments. Whether you're engaging in urban combat or fighting in tight spaces, the PL-51 Plasma Carbine is the weapon of choice for any soldier looking for a reliable and versatile plasma carbine."
```
Comment (XSlayer300) [inline code/modules/projectiles/guns/plasma.dm:184]:
```suggestion
	desc = "The lightweight plasma weapon for adaptive ranged combat. Coming with a shotgun and a burst 'tri-fire' setting for close and long engagements, in addition to the standard fully-automatic mode. Kongjian's sleek PL series of plasma-discharging weapons acts as a competitor to TE series of laser weapons used by the TGMC, allowing flexibility even in the most stressful combat situations, though the usage of highest quality materials and specifications made large-scale use and procurement all but impossible. Uses proprietary WML plasma batteries, and overheats when used too much, depending on setting."
```

File: code/modules/projectiles/guns/plasma.dm [hunk 11]
```diff
@@ -0,0 +1,385 @@
...
+	icon_state = "plasma_carbine"
+	fire_mode = GUN_FIREMODE_AUTOMATIC
+	fire_delay = 0.15 SECONDS
+	heat_per_fire = 3
+	rounds_per_shot = 25
+	radial_icon_state = "plasma_multi"
+
+/obj/item/weapon/gun/energy/lasgun/lasrifle/plasma/pistol
+	name = "\improper PL-85 plasma pistol"
+	desc = "Introducing the PL-85 Plasma Pistol, a sleek and reliable weapon for any combat situation. With its compact design and advanced features, the PL-85 is the perfect sidearm for any soldier. This powerful plasma pistol is built for speed and accuracy, allowing you to take out your enemies with ease. Its advanced targeting system and high-powered scope make it easy to acquire and track targets, while its standard and automatic fire firemodes give you the flexibility to adapt to any combat situation. Constructed from high-quality materials, the PL-85 is built to withstand even the toughest of environments. Whether you're engaged in close quarters combat or need a reliable sidearm for backup, the PL-85 Plasma Pistol is the weapon of choice for any soldier looking for a dependable and powerful plasma pistol."
```
Comment (XSlayer300) [inline code/modules/projectiles/guns/plasma.dm:258]:
```suggestion
	desc = "A highly compact pistol by energy weapon standards, the PL-85 was designed to last in desolated environments as an emergency service sidearm for non-combative personnel. Features a semi-auto, full-auto and a 'bouncy' mode setting. Kongjian's sleek PL series of plasma-discharging weapons acts as a competitor to TE series of laser weapons used by the TGMC, allowing flexibility even in the most stressful combat situations, though the usage of highest quality materials and specifications made large-scale use and procurement all but impossible. Uses proprietary WML plasma batteries, and overheats when used too much, depending on setting."
```

**Comment by XSlayer300** [comment]
screw you

_un GPTs your GPT description_

**Comment by XSlayer300** [comment]
also, *cough cough*

image

**Comment by ivanmixo** [comment]
> i don't approve this if you don't add it to the pool of eord weapon spawns

I don't approve you using dreammaker :hampter:

**Comment by Lumipharon** [comment]
Fixed a couple of bugs, and loaded a pile of basic walance tweaks, mainly for stats that were missing/wack/not in the right place.
They'll still need testing and tweaking, but this should give it a more sane baseline.

**Comment by blackdav123** [comment]
overheats are really hard to reach and magdumping with most guns wont even overheat
in the rare case you are dumping two mags the overheat only lasts for a second which barely slows down the firerate at all
on the cannon in particular the overheat is almost as long as the firedelay so there is nearly no interference from the overheat
some guns like the carbine on shotgun and pistol on standard deal ridiculous amounts of damage, with a single mag from the pistol able to kill a primo king 

for these guns to be anywhere near balanced the overheat mechanic needs to be much more punishing than a single second of shooting lost

**Comment by blackdav123** [comment]
also the pistols are fake sidearms and dont fit into pistol holsters, and they can take pistol lace as well as mag harness.

**Review by ivanmixo — CHANGES_REQUESTED** [review]
The plasma pistol is all kinds of fucked. Has 83 shots when first vended and seems to fire standard mode rounds while being full auto. It can also take attachies it shouldn't apparently.

Need to also reconsider if they should all be onehanded. And look over what attachies they can take.

**Comment by Lumipharon** [comment]
Unfugged pistol issues and some carbine issues. The 2 big guns can no longer be one handed.

**Comment by XSlayer300** [comment]
Holster sprite isn't here.

**Comment by Aporhtonoma** [comment]
shotgun machinegun is strong as all hell massive spread with massive mag and massive firerate bad
minigun can and WILL set the entire screen on fire in less than a second with a bullet to spare bad
minigun has incidiary ammo... with decent damage... bad
the sniper , rifle i dint try or dint go against that much so i dont have an oppinion on but they actually seem ok
the overheat mechanic is non existant basically.It might have come into play once when i was spamming globs with the minigun.Make it more punishing or easier to proc please.

**Comment by FosterGrime-Heymynameslouie** [comment]
the description should probably be moved to the lore information section when looking for relevant information instead of being in the description, because its a rather long description that dosent really explain what the firemodes do, and also the firemode description on the relevant information panel has nothign

**Comment by imsxz** [comment]
Hop on and play xeno against these mister coder. Carbine tri-shot and whichever one makes the big glob of fire are stupid strong. Other ones have insane damage too but those 2 are the most unreasonable.

**Comment by imsxz** [comment]
also i played a couple rounds as marine and never triggered/noticed overheat while magdumping. Have to try very hard to force overheat to trigger. Basically a fake mechanic

**Review by ivanmixo — CHANGES_REQUESTED** [review]
Reconsider tri fire mode damage/fire rate. Also the falloff and letting guns that can be one handed having tri fire in general.

**Comment by Texicos** [comment]
Carbine needs the can't fire unwielded flag

**Comment by Lumipharon** [comment]
was this good to TM again? I don't know what was happening with people updating the walance

**Comment by TiviPlus** [comment]
Apparently they did it

**Comment by TiviPlus** [comment]
i broke some sprites though

**Comment by Texicos** [comment]
when merge

**Comment by Pariah919** [comment]
someone said appreantly them being broken is my fault but i had nothing to do with this beyond yelling random 2 sentence ideas at blase

**Comment by FosterGrime-Heymynameslouie** [comment]
6 months in the making

6th times the charm

**Comment by Joe13413** [comment]
gotta love how descriptive this PR is
"adds plasma guns, they have uhhh overheat and unique fire modes" yeah bro can you tell me whats up with these new weapons?

**Comment by Victor239** [comment]
These guns were pretty insane in the test merge this weekend to the point most marines were saying "this is busted OP" and xenos "this is beyond unfun". Lore text says "Plasma weapons are rare and powerful" so this shouldn't be in the vendor for starters, maybe give them to a very rare ERT.

File: code/modules/projectiles/guns/plasma.dm [hunk 12]
```diff
@@ -0,0 +1,438 @@
...
+	flags_gun_features = GUN_CAN_POINTBLANK|GUN_AMMO_COUNTER|GUN_AMMO_COUNT_BY_SHOTS_REMAINING
+	accuracy_mult = 1
+	accuracy_mult_unwielded = 0.5
+	scatter = 4
+	scatter_unwielded = 35
+	movement_acc_penalty_mult = 5
+	damage_falloff_mult = 0.25
+
+/obj/item/weapon/gun/energy/lasgun/lasrifle/plasma/minigun
+	name = "\improper PL-73 plasma thrower"
```
Comment (XSlayer300) [inline code/modules/projectiles/guns/plasma.dm:18]:
```suggestion
	name = "\improper PL-73 plasma thrower"
```

You know, eventually we'll have to change it to TL or TX or something.

**Comment by Tristan63** [comment]
the emprah protects

**Comment by ButterChugger** [comment]
I want to believe

**Comment by Kuro020** [comment]
I still believe

**Comment by Kuro020** [comment]
image

**Comment by David-Stormwell** [comment]
Krill issue

**Pull request closed without merging**